### PR TITLE
port to 1.18

### PR DESCRIPTION
**Pull request by justliliandev**
I've tested the functionality using: https://www.curseforge.com/minecraft/texture-packs/mc-connected-ctm/files

Some variables still use old names like world instead of level
and facing instead of direction, but that has happened somewhere before 1.16 so I'm not sure if it's a problem.
I could fix it though, if requested

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -87,7 +85,7 @@ processResources {
     // this will ensure that this task is redone when the versions change.
     inputs.property 'version', project.version
     inputs.property 'mcversion', minecraft_version
-
+    duplicatesStrategy = 'include'
```
Comment (ChloeDawn) [inline build.gradle]: Why was this added?

File: src/main/java/team/chisel/ctm/client/model/AbstractCTMBakedModel.java [hunk 2]
```diff
@@ -256,39 +256,39 @@ public boolean isGui3d() {
...
+	public boolean usesBlockLight() {
+		return getParent().usesBlockLight();
 	}
 
     private <T> T applyToParent(Random rand, Function<AbstractCTMBakedModel, T> func) {
-        IBakedModel parent = getParent(rand);
-        if (parent instanceof AbstractCTMBakedModel) {
-            return func.apply((AbstractCTMBakedModel) parent);
+        BakedModel parent = getParent(rand);
+        if (parent instanceof AbstractCTMBakedModel ctmBakedModel) {
```
Comment (ChloeDawn) [inline src/main/java/team/chisel/ctm/client/model/AbstractCTMBakedModel.java:290]: There is significant overuse of pattern matching with no thought put into variable names.
If anything this is *less* readable than the previous implementation.

**Review by ChloeDawn — CHANGES_REQUESTED** [review]

File: src/main/java/team/chisel/ctm/client/util/ConnectionLocations.java [hunk 3]
```diff
@@ -104,17 +100,17 @@ private ConnectionLocations(Direction normal, @Nullable Dir dir, boolean offset)
...
         }
 
         if (offset) {
-            pos = pos.offset(normal);
+            pos = pos.relative(normal);
         }
         return pos;
     }
-
+
```
Comment (ChloeDawn) [inline src/main/java/team/chisel/ctm/client/util/ConnectionLocations.java]: Indentation on new line
```suggestion

```

File: src/main/java/team/chisel/ctm/client/util/Dir.java [hunk 4]
```diff
@@ -59,7 +58,7 @@
 	
 	private @NonnullType BlockPos[] offsets = new BlockPos[6];
 
-	private Dir(Direction... dirs) {
+	Dir(Direction... dirs) {
```
Comment (ChloeDawn) [inline src/main/java/team/chisel/ctm/client/util/Dir.java:61]: Why was this made package-private?

File: src/main/java/team/chisel/ctm/client/util/ConnectionLocations.java [hunk 5]
```diff
@@ -67,19 +67,15 @@
     private final @Nullable Dir dir;
     private boolean offset;
 
-    private ConnectionLocations(@Nullable Dir dir) {
+    ConnectionLocations(@Nullable Dir dir) {
```
Comment (ChloeDawn) [inline src/main/java/team/chisel/ctm/client/util/ConnectionLocations.java:70]: Why were these made package-private?

File: src/main/java/team/chisel/ctm/client/texture/IMetadataSectionCTM.java [hunk 6]
```diff
@@ -145,17 +145,16 @@ public static IMetadataSectionCTM fromJson(JsonObject obj) throws JsonParseExcep
...
-        public @Nullable IMetadataSectionCTM deserialize(@Nullable JsonObject json) throws JsonParseException {
+        public @Nullable IMetadataSectionCTM fromJson(@Nullable JsonObject json) throws JsonParseException {
             if (json != null && json.isJsonObject()) {
                 JsonObject obj = json.getAsJsonObject();
                 if (obj.has("ctm_version")) {
                     JsonElement version = obj.get("ctm_version");
                     if (version.isJsonPrimitive() && version.getAsJsonPrimitive().isNumber()) {
-                        switch (version.getAsInt()) {
-                        case 1:
+                        if (version.getAsInt() == 1) {
```
Comment (ChloeDawn) [inline src/main/java/team/chisel/ctm/client/texture/IMetadataSectionCTM.java]: This was a switch to make diffs cleaner for future version additions

File: src/main/java/team/chisel/ctm/client/texture/IMetadataSectionCTM.java [hunk 7]
```diff
@@ -63,9 +63,9 @@
             }
         }
         return meta.getType().makeTexture(new TextureInfo(
-                Arrays.stream(ObjectArrays.concat(sprite.getName(), meta.getAdditionalTextures()))
-                	  .map(rl -> new RenderMaterial(AtlasTexture.LOCATION_BLOCKS_TEXTURE, rl))
-                	  .map(bakedTextureGetter::apply)
+                Arrays.stream(ObjectArrays.concat(new ResourceLocation[]{sprite.getName()}, meta.getAdditionalTextures(), ResourceLocation.class))
```
Comment (ChloeDawn) [inline src/main/java/team/chisel/ctm/client/texture/IMetadataSectionCTM.java]: This is a change of semantics, as per documentation of the `ObjectArrays#concat` methods

https://guava.dev/releases/21.0/api/docs/com/google/common/collect/ObjectArrays.html#concat-T-T:A-
https://guava.dev/releases/21.0/api/docs/com/google/common/collect/ObjectArrays.html#concat-T:A-T-

File: src/main/java/team/chisel/ctm/client/model/parsing/ModelLoaderCTM.java [hunk 8]
```diff
@@ -15,7 +15,7 @@
 
 	INSTANCE;
 
-	private static final Map<Integer, IModelParser> parserVersions = ImmutableMap.of(1, new ModelParserV1());
+	private static final Map<Integer, IModelParser> parserVersions = Map.of(1, new ModelParserV1());
```
Comment (ChloeDawn) [inline src/main/java/team/chisel/ctm/client/model/parsing/ModelLoaderCTM.java:17]: This file still contains the import for Guava's ImmutableMap class

**Review by ChloeDawn — CHANGES_REQUESTED** [review]

File: src/main/java/team/chisel/ctm/client/util/ConnectionLocations.java (hunk 5 above)
Comment (PepperCode1) [reply src/main/java/team/chisel/ctm/client/util/ConnectionLocations.java:70]: Enum constructors are always private. This change was made to remove the unnecessary modifier.

File: src/main/java/team/chisel/ctm/client/util/Dir.java (hunk 4 above)
Comment (justliliandev) [reply src/main/java/team/chisel/ctm/client/util/Dir.java:61]: it's an enum, enum constructors are private by default

File: src/main/java/team/chisel/ctm/client/model/AbstractCTMBakedModel.java (hunk 2 above)
Comment (justliliandev) [reply src/main/java/team/chisel/ctm/client/model/AbstractCTMBakedModel.java:290]: do you have any ideas for a better name or should I just remove the pattern matching there?

File: build.gradle (hunk 1 above)
Comment (justliliandev) [reply build.gradle]: I've added a comment explaining this on commit #d4006fb:
duplicate strategy is necessary, but i'm not sure why. gradle complains that accesstransformer.cfg is a duplicate...

I thought that's maybe because of the new modules system but I really don't know

File: src/main/java/team/chisel/ctm/client/texture/IMetadataSectionCTM.java (hunk 6 above)
Comment (justliliandev) [reply src/main/java/team/chisel/ctm/client/texture/IMetadataSectionCTM.java]: okay, will undo that change

File: src/main/java/team/chisel/ctm/client/texture/IMetadataSectionCTM.java (hunk 7 above)
Comment (justliliandev) [reply src/main/java/team/chisel/ctm/client/texture/IMetadataSectionCTM.java]: yes it was, but it didn't want to compile for me, so I "fixed" it. I tried the previous ObjectArrays#concat call and now it works...

File: src/main/java/team/chisel/ctm/client/model/parsing/ModelLoaderCTM.java (hunk 8 above)
Comment (justliliandev) [reply src/main/java/team/chisel/ctm/client/model/parsing/ModelLoaderCTM.java:17]: removed

File: build.gradle (hunk 1 above)
Comment (justliliandev) [reply build.gradle]: I've tested it again and now it works without duplicatesStrategy, therefore I removed it

**Comment by justliliandev** [comment]
could I get another review /reavulation of the changes?

**Comment by Edivad99** [comment]
I would change all the names from `Level world` to `Level level`
